Enabling inplace in quantized relu

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23704 Enabling inline in quantized relu**

Differential Revision: [D16634539](https://our.internmc.facebook.com/intern/diff/D16634539)

Commit:
Enabling inline in quantized relu

dzhulgakov reviewed (changes requested):
This has broken tests and you can't reuse the same function as it has different signature. Also, you probably want a test for it.

Commit:
Update on "Enabling inline in quantized relu"

zafartahirov commented:
Added tests

Commit:
Update on "Enabling inline in quantized relu"


Differential Revision: [D16634539](https://our.internmc.facebook.com/intern/diff/D16634539)

jerryzh168 commented on test/test_quantized.py:
```diff
@@ -95,15 +98,24 @@ def test_qrelu(self, X):
             self.assertEqual(qY, qY_hat, message="{} relu failed".format(name))
 
+        ops_under_test_inline = {
+            'inline native': torch.relu_,
```
you mean `inplace`?

jerryzh168 reviewed (approved):
lgtm, please fix the naming

Commit:
Update on "Enabling inline in quantized relu"


Differential Revision: [D16634539](https://our.internmc.facebook.com/intern/diff/D16634539)

Commit:
Update on "Enabling inline in quantized relu"


Differential Revision: [D16634539](https://our.internmc.facebook.com/intern/diff/D16634539)

Commit:
Update on "Enabling inline in quantized relu"


Differential Revision: [D16634539](https://our.internmc.facebook.com/intern/diff/D16634539)

ezyang commented:
I see a bunch of test failures on master after this commit:

```
Aug 09 13:40:43 ======================================================================
Aug 09 13:40:43 ERROR: test_skip_quant (__main__.PostTrainingQuantTest)
Aug 09 13:40:43 The case when we want to skip quantizing some layers
Aug 09 13:40:43 ----------------------------------------------------------------------
Aug 09 13:40:43 Traceback (most recent call last):
Aug 09 13:40:43   File "test_quantization.py", line 212, in test_skip_quant
Aug 09 13:40:43     checkQuantized(model)
Aug 09 13:40:43   File "test_quantization.py", line 210, in checkQuantized
Aug 09 13:40:43     self.checkScriptable(model, self.calib_data)
Aug 09 13:40:43   File "/var/lib/jenkins/workspace/test/common_quantization.py", line 123, in checkScriptable
Aug 09 13:40:43     scripted = torch.jit.script(orig_mod)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1169, in script
Aug 09 13:40:43     return _convert_to_script_module(obj)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1842, in _convert_to_script_module
Aug 09 13:40:43     return WeakScriptModuleProxy(mod, stubs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1416, in init_then_register
Aug 09 13:40:43     original_init(self, *args, **kwargs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1766, in __init__
Aug 09 13:40:43     _create_methods_from_stubs(self, stubs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1377, in _create_methods_from_stubs
Aug 09 13:40:43     self._c._create_methods(self, defs, rcbs, defaults)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 981, in _make_strong_submodule
Aug 09 13:40:43     new_strong_submodule = _convert_to_script_module(module)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1842, in _convert_to_script_module
Aug 09 13:40:43     return WeakScriptModuleProxy(mod, stubs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1416, in init_then_register
Aug 09 13:40:43     original_init(self, *args, **kwargs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1766, in __init__
Aug 09 13:40:43     _create_methods_from_stubs(self, stubs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1377, in _create_methods_from_stubs
Aug 09 13:40:43     self._c._create_methods(self, defs, rcbs, defaults)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 981, in _make_strong_submodule
Aug 09 13:40:43     new_strong_submodule = _convert_to_script_module(module)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1842, in _convert_to_script_module
Aug 09 13:40:43     return WeakScriptModuleProxy(mod, stubs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1416, in init_then_register
Aug 09 13:40:43     original_init(self, *args, **kwargs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1766, in __init__
Aug 09 13:40:43     _create_methods_from_stubs(self, stubs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1377, in _create_methods_from_stubs
Aug 09 13:40:43     self._c._create_methods(self, defs, rcbs, defaults)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 981, in _make_strong_submodule
Aug 09 13:40:43     new_strong_submodule = _convert_to_script_module(module)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1842, in _convert_to_script_module
Aug 09 13:40:43     return WeakScriptModuleProxy(mod, stubs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1416, in init_then_register
Aug 09 13:40:43     original_init(self, *args, **kwargs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1766, in __init__
Aug 09 13:40:43     _create_methods_from_stubs(self, stubs)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1377, in _create_methods_from_stubs
Aug 09 13:40:43     self._c._create_methods(self, defs, rcbs, defaults)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1008, in _try_compile_fn
Aug 09 13:40:43     return torch.jit.script(fn, _rcb=rcb)
Aug 09 13:40:43   File "/opt/conda/lib/python3.6/site-packages/torch/jit/__init__.py", line 1192, in script
Aug 09 13:40:43     fn = torch._C._jit_script_compile(qualified_name, ast, _rcb, get_default_args(obj))
Aug 09 13:40:43 RuntimeError: 
Aug 09 13:40:43 builtin cannot be used as a value:
Aug 09 13:40:43 at /opt/conda/lib/python3.6/site-packages/torch/nn/quantized/functional.py:19:12
Aug 09 13:40:43 def relu(input, inplace=False):
Aug 09 13:40:43     # type: (Tensor, bool) -> Tensor
Aug 09 13:40:43     r"""relu(input, inplace=False) -> Tensor
Aug 09 13:40:43 
Aug 09 13:40:43     Applies the rectified linear unit function element-wise. See
Aug 09 13:40:43     :class:`~torch.nn.ReLU` for more details.
Aug 09 13:40:43     """
Aug 09 13:40:43     if not input.is_quantized:
Aug 09 13:40:43            ~~~~~~~~ <--- HERE
Aug 09 13:40:43         raise ValueError("Input to 'quantized.relu' must be quantized!")
Aug 09 13:40:43     if inplace:
Aug 09 13:40:43         return torch.relu_(input)
Aug 09 13:40:43     else:
Aug 09 13:40:43         return torch.relu(input)
Aug 09 13:40:43 
Aug 09 13:40:43 'relu' is being compiled since it was called from '__module__.__torch__.torch.nn.quantized.modules.activation.ReLU.forward'
Aug 09 13:40:43 at /opt/conda/lib/python3.6/site-packages/torch/nn/quantized/modules/activation.py:36:16
Aug 09 13:40:43     def forward(self, input):
Aug 09 13:40:43         return F.relu(input)
Aug 09 13:40:43                ~~~~~~~~~~~~ <--- HERE
Aug 09 13:40:43 
Aug 09 13:40:43 '__module__.__torch__.torch.nn.quantized.modules.activation.ReLU.forward' is being compiled since it was called from '__module__.__torch__.common_quantization.InnerModule.forward'
Aug 09 13:40:43 at /var/lib/jenkins/workspace/test/common_quantization.py:284:16
Aug 09 13:40:43     def forward(self, x):
Aug 09 13:40:43         return self.relu(self.fc2(self.relu(self.fc1(x))))
Aug 09 13:40:43                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
Aug 09 13:40:43 
Aug 09 13:40:43 '__module__.__torch__.common_quantization.InnerModule.forward' is being compiled since it was called from '__module__.__torch__.torch.quantization.quantize.___torch_mangle_109.QuantWrapper.forward'
Aug 09 13:40:43 at /opt/conda/lib/python3.6/site-packages/torch/quantization/quantize.py:104:9
Aug 09 13:40:43     def forward(self, X):
Aug 09 13:40:43         X = self.quant(X)
Aug 09 13:40:43         X = self.module(X)
Aug 09 13:40:43         ~~~~~~~~~~~~~~~~~ <--- HERE
Aug 09 13:40:43         return self.dequant(X)
Aug 09 13:40:43 
Aug 09 13:40:43 '__module__.__torch__.torch.quantization.quantize.___torch_mangle_109.QuantWrapper.forward' is being compiled since it was called from '__module__.__torch__.common_quantization.SkipQuantModel.forward'
Aug 09 13:40:43 at /var/lib/jenkins/workspace/test/common_quantization.py:299:24
Aug 09 13:40:43     def forward(self, x):
Aug 09 13:40:43         return self.fc(self.sub(x))
Aug 09 13:40:43                        ~~~~~~~~~~ <--- HERE
Aug 09 13:40:43 
Aug 09 13:40:43 
Aug 09 13:40:43 
Aug 09 13:40:43 ----------------------------------------------------------------------
Aug 09 13:40:43 Ran 12 tests in 1.022s
Aug 09 13:40:43 
Aug 09 13:40:43 FAILED (errors=1)
Aug 09 13:40:43 Traceback (most recent call last):
Aug 09 13:40:43   File "test/run_test.py", line 416, in <module>
Aug 09 13:40:43     main()
Aug 09 13:40:43   File "test/run_test.py", line 408, in main
Aug 09 13:40:43     raise RuntimeError(message)
Aug 09 13:40:43 RuntimeError: test_quantization failed!
Aug 09 13:40:44 + cleanup
Aug 09 13:40:44 + retcode=1
Aug 09 13:40:44 + set +x
```

jamesr66a commented:
I think this PR was likely based before the commit to enable OSS testing and that's why signal for that breakage^ didn't show up in CI/sandcastle

jamesr66a commented:
Fix here https://github.com/pytorch/pytorch/pull/24099

zafartahirov commented:
So is this passing the tests, or anything else needs to be done for itt?

zafartahirov commented:
Wait, why is this reverted? Wasn't the fix submitted?

ezyang commented:
> Wait, why is this reverted? Wasn't the fix submitted?

It is (a lot) quicker to unland, than it is to wait for fixes
